### PR TITLE
Refactors PM UI code to reduce duplicated calls to fetch metadata

### PR DESCRIPTION
**Pull request by dominoFire**
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/1639

Regression? Last working version:

## Description

- Refactors PM UI code to isolate package metadata operation to stay in `PackageItemViewModel` (PIVM) class; related objects just call PIVM method named `ReloadPackageMetadataAsync(NuGetVersion newVersion, CancellationToken token)` to fetch metadata from search service. 
- Adds an additional member to PIVM called `PackageMetadata` to cache existing metadata.
- Removes lazy metadata in PIVM to avoid code duplication.
- Adds cancellation tokens where needed.
- Adds tests.
- Removes duplicated code found in multiple places.
- Fixes some binding errors found when refactoring.
- Addressed Code Analyzer suppressions as needed.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added: new unit tests on added cancellation tokens
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A: No major product changes

### CODE REVIEW

**Review by jeffkl — APPROVED** [review]

**Comment by erdembayar** [comment]
@dominoFire 
Code change is quite large. Is it possible to add Github comments as what being changed in each file?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDetailControlModel.cs [hunk 1]
```diff
@@ -28,15 +28,31 @@ public class PackageDetailControlModel : DetailControlModel
             IEnumerable<IProjectContextInfo> projects)
             : base(serviceBroker, projects)
         {
+            if (solutionManager == null)
+            {
+                throw new ArgumentNullException(nameof(solutionManager));
+            }
+
             _solutionManager = solutionManager;
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDetailControlModel.cs:36]:
```suggestion
            _solutionManager = solutionManager ?? throw new ArgumentNullException(nameof(solutionManager));
```

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs [hunk 2]
```diff
@@ -34,19 +34,34 @@ public abstract class DetailControlModel : INotifyPropertyChanged, IDisposable
         protected IEnumerable<IProjectContextInfo> _nugetProjects;
 
         // all versions of the _searchResultPackage
-        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1051:DoNotDeclareVisibleInstanceFields")]
-        protected List<(NuGetVersion version, bool isDeprecated)> _allPackageVersions;
+        private protected List<(NuGetVersion version, bool isDeprecated)> _allPackageVersions;
 
-        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1051:DoNotDeclareVisibleInstanceFields")]
-        protected PackageItemViewModel _searchResultPackage;
+        private protected PackageItemViewModel _searchResultPackage;
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs:39]: Why `private protected`?  This is something that I never remember the semantic meaning of, so I always have to keep [looking it up in the docs](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/private-protected). It means "A private protected member is accessible by types derived from the containing class, but only within its containing assembly."

Is this abstract class derived by any class in a different assembly? If not, then semantically `protected` is the same, and the `private` accessor isn't used.

Or was it just to suppress CA1051?  In which case perhaps making the field a property instead might be better since it won't confuse people like me about what `private protected` means and trying to guess why this ununual access level is being used.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs [hunk 3]
```diff
@@ -483,46 +424,26 @@ private static class PackageDeprecationReason
...
-                    }
-
-                    PackageDeprecationReasons = newDeprecationReasons;
-                    PackageDeprecationAlternatePackageText = newAlternatePackageText;
-
                     PackageVulnerabilities = _packageMetadata?.Vulnerabilities?.ToList();
 
                     OnPropertyChanged(nameof(PackageMetadata));
+
+                    OnPropertyChanged(nameof(PackageDeprecationReasons));
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs:439]: nitpick: unnecessary whitespace line

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs [hunk 4]
```diff
@@ -884,5 +780,36 @@ protected void SetAutoReferencedCheck(bool autoReferenced)
...
+        }
+
+        private async Task RecreateVersionsAsync(CancellationToken token)
+        {
+            token.ThrowIfCancellationRequested();
+
+            await CreateVersionsAsync(token);
+
+            NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(async () => await OnCurrentPackageChangedAsync(token))
+                .PostOnFailure(nameof(DetailControlModel), nameof(OnCurrentPackageChangedAsync));
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs:812]: This needs to handle the `OperationCancelledException` when `token` is requesting cancellation.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs [hunk 5]
```diff
@@ -601,19 +522,20 @@ public DisplayVersion SelectedVersion
...
                             {
-                                // Expected
-                            }
-                        })
-                            .PostOnFailure(nameof(DetailControlModel));
+                                try
+                                {
+                                    await SelectedVersionChangedAsync(_searchResultPackage, _selectedVersion.Version, loadCts.Token);
+                                }
+                                catch (OperationCanceledException) when (loadCts.IsCancellationRequested)
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs:533]: I know you didn't write this, but looking at it now, it makes me worried about scoping rules. If `loadCts` is cancelled and then replaced with a new instance, then the new `loadCts` will not be requesting cancellation, and therefore I don't know if `loadCts.IsCancellationRequested` will be true or false.

I think if you do `var cancellationToken = loadCts.Token;` before `JTF.RunAsync`, and use that in the two places (most importantly the catch's when clause), then that should guarantee that the correct token is captured.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs [hunk 6]
```diff
@@ -884,5 +780,36 @@ protected void SetAutoReferencedCheck(bool autoReferenced)
...
+                PackageDeprecationMetadataContextInfo deprecation;
+                (meta, deprecation) = await searchResultPackage.ReloadPackageMetadataAsync(newVersion, cancellationToken);
+
+                result = new DetailedPackageMetadata(meta, deprecation, meta?.DownloadCount);
+            }
+
+            return result;
+        }
+
+        private async Task RecreateVersionsAsync(CancellationToken token)
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs:805]: nitpick: `CancellationTokens` are always named `cancellationToken` throughout VS and .NET. See also https://github.com/NuGet/NuGet.Client/pull/4471#discussion_r813373026

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/DetailControl.xaml.cs [hunk 7]
```diff
@@ -95,48 +94,25 @@ public void Refresh()
...
                 // once execution reaches here and thus 'model' could be null.
-                var model = DataContext as DetailControlModel;
-
-                if (model != null)
+                if (DataContext is DetailControlModel model)
                 {
                     await model.RefreshAsync(CancellationToken.None);
                 }
-            }).PostOnFailure(nameof(DetailControl));
+            }).PostOnFailure(nameof(DetailControl), nameof(Refresh));
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/DetailControl.xaml.cs:101]: nitpick: `PostOnFailure` specifies the second argument with the `[CallingMemberName]` attribute, and therefore `nameof(Refresh)` is automatically added by the compiler.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Models/V3DetailControlModelTests.cs [hunk 8]
```diff
@@ -24,18 +24,20 @@
 using Xunit;
 using Task = System.Threading.Tasks.Task;
 
-namespace NuGet.PackageManagement.UI.Test.Models
+namespace NuGet.PackageManagement.UI.Test
```
Comment (zivkan) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Models/V3DetailControlModelTests.cs:27]: Why the change in namespace?  it's still in the `Models` directory, and it helps group tests together in Test Explorer. I find it useful when tests fail to be expand to the group (namespace) with a failing test, so it doesn't show hundreds of classes, only a few.

**Comment by dominoFire** [comment]
Closing to address feedback on separate PRs:

- Handling cancelation tokens
- Reducing repeated / unused code.
